_episodes/05-Dimuonspectrum.md

```diff
@@ -196,28 +206,98 @@ $$p_{\rm y parent} = p_{\rm y child 0} + p_{\rm y child 1} + p_{\rm y child 2} +
+        etot += e[i]                           # px, py, and pz components
+        pxtot += px[i]
+        pytot += py[i]
+        pztot += pz[i]                           
+        m2 = etot**2 - (pxtot**2 + pytot**2 + pztot**2)      # uses the total energy,px,py,and pz to calculate invariant mass
+    return(np.sqrt(abs(m2)))
+```
 
 ```python
-e=ak.unflatten(e,2)
-px=ak.unflatten(px,2)
-py=ak.unflatten(py,2)
-pz=ak.unflatten(pz,2)
 
+pp=[] # postive positive
+nn=[] # negative negative
+pm=[] # opposite charges
+M =[] # all combinations
+
+for i in range(0,len(q)-1,2) :  # loop every 2 muons
+    # Make a list with information for 2 muons 
+    E=[e[i],e[i+1]]
+    PX=[px[i],px[i+1]]
+    PY=[py[i],py[i+1]]
+    PZ=[pz[i],pz[i+1]]
+    M.append(invmass(E,PX,PY,PZ))
+    if q[i]*q[i+1] < 0 :
+        pm.append(invmass(E,PX,PY,PZ))
+    elif q[i]+q[i+1] == 2:
+        pp.append(invmass(E,PX,PY,PZ))
+    elif q[i]+q[i+1] == -2:
+        nn.append(invmass(E,PX,PY,PZ))
+    else : 
+        print('anomaly?')
+print("Done!")
 ```
 
-Now lets sum the 4-momenta components and calculate the event mass
+## Now we plot for all combinations
+
+I would like you to make these 4 histograms *in 4 different ways* focusing on on different mass ranges. To look at these mass ranges, you'll want to use the `bins` and `range` options in the `plt.hist()` function.
+
+- *Mass range 1*: 0 - 120
+- *Mass range 2*: 2 - 4
+- *Mass range 3*: 8 - 12
+- *Mass range 4*: 70 - 120
+
+Remember, you'll have 4 charge combinations for each of these histograms.
+
+1. All charge combinations
+2. Only positve muons
+3. Only negative muons
+4. Only oppositly charged muons
+
+Below I will give you some code to get you started. Please make your changes/additions below this cell and look at each mass range.
 
 ```python
-# we use axis = 1 because it squishes the row dimension and gives us 1 value per row (i.e. 1 value per event)
-e = np.sum(e,axis=1)
-px = np.sum(px,axis=1)
-py = np.sum(py,axis=1)
-pz = np.sum(pz,axis=1)
+# mass = 0-120
+
+plt.figure(figsize=(16,10))
+plt.subplot(221)
+plt.hist(M,bins=100,range=(0,120),histtype='step',label='All charge combinations')
+plt.xlabel(r'Mass (GeV/c$^2$)',fontsize=14)
+plt.legend(fontsize=18)
+
+
+plt.subplot(222)
+plt.hist(pp,bins=100,range=(0,120),histtype='step',label='$2+$')
+plt.xlabel(r'Mass (GeV/c$^2$)',fontsize=14)
+plt.legend(fontsize=18)
+
+
+plt.subplot(223)
+plt.hist(nn,bins=100,range=(0,120),histtype='step',label='$2-$')
+plt.xlabel(r'Mass (GeV/c$^2$)',fontsize=14)
+plt.legend(fontsize=18)
+
+
+plt.subplot(224)
+plt.hist(pm,bins=100,range=(0,120),histtype='step',label='Electrically neutral')
+plt.xlabel(r'Mass (GeV/c$^2$)',fontsize=14)
+plt.legend(fontsize=18)
+
+plt.tight_layout();
 ```
 
-# Exercise: Now calculate the mass per event and make the plot.
+## Exercise: Now calculate the mass per event and make the plot.
 
 > ## Solution
```

Comment:
There is only 2 things I would suggest. And that is remove the `import awkward as ak` since it's not needed here (if it's in the notebook then it's my bad) 
and Also the solution shown doesn't really apply anymore as it is written.

It could be a cp paste of the suggested code just changing a few of the ranges but also I would like to change the binning to be
`logbins=np.logspace(0,2.5,200)`
and then in the `plt.hist()` method one can do `bins=logbins`
This makes the bins linearly spaced when plotting the x-axis in log scale. 

and hence the solution image is slightly different as well